src/frontend/src/datafusion/iceberg_executor.rs

```diff
@@ -113,135 +120,201 @@ impl ExecutionPlan for IcebergScan {
 }
 
 impl IcebergScan {
-    pub fn new(
+    pub async fn new(
         provider: &IcebergTableProvider,
         // TODO: handle these params
         _projection: Option<&Vec<usize>>,
         _filters: &[Expr],
         _limit: Option<usize>,
     ) -> DFResult<Self> {
-        if provider.iceberg_scan_type != IcebergScanType::DataScan {
-            return Err(DataFusionError::NotImplemented(
-                "Only DataScan is supported currently".to_owned(),
-            ));
-        }
-
         let plan_properties = PlanProperties::new(
             EquivalenceProperties::new(provider.schema()),
             // TODO: determine partitioning
             Partitioning::UnknownPartitioning(1),
             EmissionType::Incremental,
             Boundedness::Bounded,
         );
-        let column_names = provider
+        let mut column_names: Vec<String> = provider
             .arrow_schema
             .fields()
             .iter()
             .map(|f| f.name().clone())
             .collect();
+        let table = provider
+            .iceberg_properties
+            .load_table()
+            .await
+            .map_err(to_datafusion_error)?;
+        let need_seq_num = column_names
+            .iter()
+            .any(|name| name == ICEBERG_SEQUENCE_NUM_COLUMN_NAME);
+        let need_file_path_and_pos = column_names
+            .iter()
+            .any(|name| name == ICEBERG_FILE_PATH_COLUMN_NAME)
+            && matches!(provider.iceberg_scan_type, IcebergScanType::DataScan);
+        column_names.retain(|name| {
+            ![
+                ICEBERG_FILE_PATH_COLUMN_NAME,
+                ICEBERG_SEQUENCE_NUM_COLUMN_NAME,
+                ICEBERG_FILE_POS_COLUMN_NAME,
+            ]
+            .contains(&name.as_str())
+        });
 
-        let inner = IcebergScanInner {
-            iceberg_properties: provider.iceberg_properties.clone(),
+        let mut inner = IcebergScanInner {
+            table,
             snapshot_id: provider.snapshot_id,
+            tasks: vec![],
             iceberg_scan_type: provider.iceberg_scan_type,
             arrow_schema: provider.arrow_schema.clone(),
-            column_names: Some(column_names),
-            need_seq_num: false,
-            need_file_path_and_pos: false,
+            column_names,
+            need_seq_num,
+            need_file_path_and_pos,
             plan_properties,
         };
+        inner.tasks = inner.list_iceberg_scan_task().try_collect().await?;
+        inner.plan_properties.partitioning = Partitioning::UnknownPartitioning(inner.tasks.len());
```

Comment:
We have a session variable called `batch_parallelism`, I think we can use this value to control the parallelism instead of using the task number directly, because the task number could be very large.